pyccel/parser/syntactic.py

```diff
@@ -633,54 +642,110 @@ def _visit_FunctionDef(self, stmt):
         if 'allow_negative_index' in decorators:
             decorators['allow_negative_index'] = tuple(str(a) for a in decorators['allow_negative_index'].args)
 
-        # extract the types to construct a header
-        if 'types' in decorators:
+        # extract the templates
+        if 'template' in decorators:
             types = []
-            results = []
-            container = types
             i = 0
-            ls = decorators['types'].args
-            while i<len(ls) :
-                arg = ls[i]
-
-                if isinstance(arg, Symbol):
-                    arg = arg.name
-                    container.append(arg)
-                elif isinstance(arg, String):
-                    arg = str(arg)
-                    arg = arg.strip("'").strip('"')
-                    container.append(arg)
-                elif isinstance(arg, ValuedArgument):
-                    arg_name = arg.name
-                    arg  = arg.value
-                    container = results
-                    if not arg_name == 'results':
-                        msg = 'Argument "{}" provided to the types decorator is not valid'.format(arg_name)
+
+            if not isinstance(decorators['template'], list):
+                decorators['template'] = [decorators['template']]
+            for comb_types in decorators['template']:
+                if len(comb_types.args) != 2:
+                        msg = 'Number of Arguments provided to the types decorator is not valid'
+                        errors.report(msg,
+                                        symbol = comb_types,
+                                        bounding_box = (stmt.lineno, stmt.col_offset),
+                                        severity='error')
+
+                tp_name = comb_types.args[0]
+                ls = comb_types.args[1]
+                if isinstance(ls, ValuedArgument):
+                    arg_name = ls.name
+                    ls  = ls.value
+                    if not arg_name == 'types':
```

Comment:
We force the `name` arg to never be a keyword arg?

Reply:
`@template('z', types=['int', 'real'])`, we are giving the choice for the second argument to be a keyword arg or not while forcing it the key to be `types` in case of keyword arg

Reply:
So yes?

Reply:
Why?

Reply:
I am not sure where the problem exactly is, do you mean we should force the list of types to be a keyword arg?

Reply:
No I mean you have said that I am not allowed to do:
`@template(name='z', types=['int', 'real']`
But I don't understand why

Reply:
yes you are right this case is not handled properly

Reply:
it's now fixed